src/compiler/checker.ts

```diff
@@ -3309,7 +3309,18 @@ namespace ts {
                     const tsExtension = tryExtractTSExtension(moduleReference);
                     if (tsExtension) {
                         const diag = Diagnostics.An_import_path_cannot_end_with_a_0_extension_Consider_importing_1_instead;
-                        error(errorNode, diag, tsExtension, removeExtension(moduleReference, tsExtension));
+                        const importSourceWithoutExtension = removeExtension(moduleReference, tsExtension);
+                        let replacedImportSource = importSourceWithoutExtension;
+                        /**
+                         * Direct users to import source with .js extension if outputting an ES module.
+                         * @see https://github.com/microsoft/TypeScript/issues/42151
+                         */
+                        const target = getEmitScriptTarget(compilerOptions);
+                        const moduleKind = getEmitModuleKind(compilerOptions);
+                        if (target >= ScriptTarget.ES2015 && moduleKind >= ModuleKind.ES2015) {

```

Comment:
I don’t think the script target is relevant independently of the module target.

Reply:
Updated this to only depend on `moduleKind`.